src/lib/kimyi.rs

```diff
@@ -0,0 +1,737 @@
+                    // We follow Corcheulo et al.'s suggestions and never follow Deletes with
+                    // Inserts.
+                },
+                _ => {
+                    r3is(parser, &dist, &n, &mut nbrs);
+                    r3ir(parser, &n, &mut nbrs);
+                }
+            }
+            r3d(parser, &n, &mut nbrs);
+            r3s_n(parser, &n, &mut nbrs);
+            let v = nbrs.into_iter()
+                        .map(|x| {
+                                let t = x.cf - n.cf;
+                                (x, t)
+                             })
+                        .collect::<Vec<(PathFNode, _)>>();
+            v
+        },
+        |n| n.cg,
+        |n| {
+            // Is n a success node?
+
+            // As presented in both Corchuelo et al. and Kim Yi, one type of success is if N
+            // symbols are parsed in one go. Indeed, without such a check, the search space quickly
+            // becomes too big. There isn't a way of encoding this check in r3s_n, so we check
+            // instead for its result: if the last N ('PARSE_AT_LEAST' in this library) repairs are
+            // shifts, then we've found a success node.
+            if n.repairs.len() > PARSE_AT_LEAST {
+                let mut all_shfts = true;
+                for x in n.repairs.vals().take(PARSE_AT_LEAST) {
+                    if let ParseRepair::Shift = *x {
+                        continue;
+                    }
+                    all_shfts = false;
+                    break;
+                }
+                if all_shfts {
+                    return true;
+                }
+            }
+
+            let (_, la_term) = parser.next_lexeme(None, n.la_idx);
+            match parser.stable.action(*n.pstack.val().unwrap(), la_term) {
+                Some(Action::Accept) => true,
+                _ => false,
+            }
+        });
+
+    if astar_opt.is_none() {
+        return (in_la_idx, vec![]);
+    }
+
+    let full_rprs = collect_repairs(astar_opt.unwrap().0);
+    let smpl_rprs = simplify_repairs(parser, full_rprs);
+    let (la_idx, mut rpr_pstack) = apply_repairs(parser,
+                                                 in_la_idx,
+                                                 start_cactus_pstack,
+                                                 &mut Some(&mut tstack),
+                                                 &smpl_rprs);
+
+    in_pstack.clear();
+    while !rpr_pstack.is_empty() {
+        let p = rpr_pstack.parent().unwrap();
+        in_pstack.push(rpr_pstack.try_unwrap()
+                                 .unwrap_or_else(|c| c.val()
+                                                      .unwrap()
+                                                      .clone()));
+        rpr_pstack = p;
+    }
+    in_pstack.reverse();
+
+    (la_idx, vec![smpl_rprs])
+}
+
+fn r3is<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
```

Comment:
I guess I'll have to read the paper to know what `r3is`, `r3ir`, `r3d`, etc, are. I guess these are fixed functions from the paper so we can't give them sensible names. But maybe their function can be summarised in a small comment?

Reply:
I think people have to read the paper to understand those... I could tell them what page to read though.